[clang][analyzer] Stable order for SymbolRef-keyed containers

necto opened the pull request:
Generalize the `SymbolID`s used for `SymbolData` to all `SymExpr`s and use these IDs for comparison `SymbolRef` keys in various containers, such as `ConstraintMap`. These IDs are superior to raw pointer values because they are more controllable and are not randomized across executions (unlike [pointers](https://en.wikipedia.org/wiki/Address_space_layout_randomization)).

These IDs order is stable across runs because SymExprs are allocated in the same order.

Stability of the constraint order is important for the stability of the analyzer results. I evaluated this change on a set of 200+ open-source C and C++ projects with the total number of ~78 000 symbolic-execution issues passing Z3 refutation.

This patch reduced the run-to-run churn (flakiness) in SE issues from 80-90 to 30-40 (out of 78K) in our CSA deployment (in our setting flaky issues are mostly due to Z3 refutation instability).

Note, most of the issue churn (flakiness) is caused by the mentioned Z3 refutation. With Z3 refutation disabled, issue churn goes down to ~10 issues out of 83K and this patch has no effect on appearing/disappearing issues between runs. It however, seems to reduce the volatility of the execution flow: before we had 40-80 issues with changed execution flow, after - 10-30.

Importantly, this change is necessary for the next step in stabilizing analysis results by caching Z3 query outcomes between analysis runs (work in progress).

Across our admittedly noisy CI runs, I detected no significant effect on memory footprint or analysis time.

This change set supersedes #121347 

CPP-5919

necto committed:
Add SymbolID to every SymExpr, not just SymbolData

necto committed:
[NFC] Group allocation and ID increment into SymExprAllocator

This helps with two things:
- make sure NextSymbolID is incremented on every allocation
- No SymExpr can be allocated without a unique ID

necto committed:
Use getSymbolID instead of raw pointer values for SymbolRef ordering

necto commented:
@steakhal @NagyDonat, I revised and improved #121347.
To facilitate review, I broke the change down into 3 self-contained commits. let tests pass after each of them.
I re-evaluated this version and discovered no difference in the analysis outcome.

steakhal reviewed (commented):
@Xazax-hun I'm pretty happy with this PR, but I'm curious about your opinion on adding the `SymbolID Sym` member to the SymExprs.

Xazax-hun reviewed (approved):
I am OK with this change. Having more stable analysis results is valuable.  

That being said, do we know how exactly the unstable addresses result in unstable results?

NagyDonat commented:
> That being said, do we know how exactly the unstable addresses result in unstable results?

I cannot pinpoint the responsible code fragment, but in general this is a typical property of associative data structures. A typical associative map doesn't preserve the order of insertion, and instead stores the key-value pairs in some arbitrary order that facilitates quick and effective access (for example, many implementations sort them according to the keys -- or the hash values of the keys). When the program iterates over a data structure of this kind, it will see the key-value pairs in this arbitrary order, which may depend on pointer values, whose order relationships are undefined and may change if the program is reran with the same input. If the calculations performed by the loop are non-commutative (e.g. "sum all the values" is commutative, "return the first value which is larger than 10" is not), then this unstable iteration order produces unstable output.

By the way, the Python language (where performance is not prioritized) rules out these subtle bugs by recording the order of insertion within its `dict` (associative map) datatype and using that as the iteration order.

necto commented:
> That being said, do we know how exactly the unstable addresses result in unstable results?

One mechanism that we studied in the past weeks is the SMT issue refutation: assertion order is that of the constraint order in the `ConstraintMap`, and it turns out Z3 runs a query differently depending on the order of the assertions.

There are likely other mechanisms, for example, the SymbolRef `EquivalenceClass` in RangeConstraintManager is an `ImmutableMap<SymbolRef, ...>`. With different ordering, the constraint manager might choose a different representative `SymbolRef` for a given class. This is speculative, though.

> By the way, the Python language (where performance is not prioritized) rules out these subtle bugs by recording the order of insertion within its `dict` (associative map) datatype and using that as the iteration order.

Interesting parallel, as this patch is applying the same strategy for SymbolRef maps

NagyDonat reviewed (approved):
LGTM, thanks for this improvement. Using the "raw" ordering of pointers is always a red flag, I strongly support eliminating it if possible.

By the way, what fraction of the results is perturbed (replaced with other random results) when this commit is activated? Did you run any "without this commit vs with this commit" comparisons?

As a tangential remark, I noticed that there seems to be heavy code duplication between the `getXSymbol` methods of `SymbolManager` -- it would be nice to unify them into a single template method that takes the symbol class and the argument list types as template parameters (analogously to `SymExprAllocator::make` which does similar forwarding). However, this probably belongs to a separate NFC refactoring commit (and I can also implement it if you don't have time for it). What do you think? Do you see any obstacle (or perhaps conflict with changes that you're planning).

necto committed:
[NFC] Explain why isDataEqual is necessary; inline type aliases.

necto commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/SymbolManager.h:
```diff
@@ -687,4 +711,35 @@ class SymbolVisitor {
+  using value_type =
+      typename ImutProfileInfo<clang::ento::SymbolRef>::value_type;
+  using value_type_ref =
+      typename ImutProfileInfo<clang::ento::SymbolRef>::value_type_ref;
```
inlined
d32f6abb0404 [NFC] Explain why isDataEqual is necessary; inline type aliases.

necto commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/SymbolManager.h:
```diff
@@ -687,4 +711,35 @@ class SymbolVisitor {
+
+  static bool isEqual(key_type_ref LHS, key_type_ref RHS) {
+    return LHS->getSymbolID() == RHS->getSymbolID();
+  }
```
Inlined
d32f6abb0404 [NFC] Explain why isDataEqual is necessary; inline type aliases.

necto commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/SymbolManager.h:
```diff
@@ -687,4 +711,35 @@ class SymbolVisitor {
+      typename ImutProfileInfo<clang::ento::SymbolRef>::value_type_ref;
+  using key_type = value_type;
+  using key_type_ref = value_type_ref;
+  using data_type = bool;
+  using data_type_ref = bool;
```
This is due to the way ImmutableMap and ImmutableSet are implemented via a common interface of AVLTree
d32f6abb0404 [NFC] Explain why isDataEqual is necessary; inline type aliases.

necto committed:
[NFC] A doc comment explaining the SymExpr::getSymbolID()

necto commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/SymExpr.h:
```diff
@@ -56,6 +68,8 @@ class SymExpr : public llvm::FoldingSetNode {
 
   Kind getKind() const { return K; }
 
+  SymbolID getSymbolID() const { return Sym; }
```
Documented in
355c588f4469 az/order-by-symref-id [NFC] A doc comment explaining the SymExpr::getSymbolID()

necto commented:
> By the way, what fraction of the results is perturbed (replaced with other random results) when this commit is activated? Did you run any "without this commit vs with this commit" comparisons?

The difference is around 5-10 appearing/disappearing issues, and 50-100 issues with changed execution path (out of 83 K issues, with no Z3 refutation).

> As a tangential remark, I noticed that there seems to be heavy code duplication between the `getXSymbol` methods of `SymbolManager` -- it would be nice to unify them into a single template method that takes the symbol class and the argument list types as template parameters (analogously to `SymExprAllocator::make` which does similar forwarding). However, this probably belongs to a separate NFC refactoring commit (and I can also implement it if you don't have time for it). What do you think? Do you see any obstacle (or perhaps conflict with changes that you're planning).

It makes sense and I have no future developments that would be incompatible with this refactoring. I did not do it in this PR to avoid scope creep. I will see on Monday, if I have some spare time to do this refactoring.

@steakhal, I believe I addressed all your feedback. If you are happy with the PR, feel free to merge it.

steakhal reviewed (approved):
LGTM, thanks.

steakhal commented:
@necto Please have a look at the build failures. There are plenty.
For the time being, I reverted this PR.
